pkg/encryption/decrypt_store.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package encryption
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"golang.org/x/crypto/sha3"
+)
+
+type decryptingStore struct {
+	storage.Getter
+}
+
+func NewDecryptingStore(s storage.Getter) storage.Getter {
+	return &decryptingStore{s}
+}
+
+func (s *decryptingStore) Get(ctx context.Context, mode storage.ModeGet, addr swarm.Address) (ch swarm.Chunk, err error) {
+	switch l := len(addr.Bytes()); l {
+	case 32:
+		// normal, unencrypted content
+		return s.Getter.Get(ctx, mode, addr)
+
+	case 64:
+		// encrypted reference
+		ref := addr.Bytes()
+		address := swarm.NewAddress(ref[:32])
+		ch, err := s.Getter.Get(ctx, mode, address)
+		if err != nil {
+			return nil, err
+		}
+
+		decryptionKey := make([]byte, KeyLength)
+		copy(decryptionKey, ref[32:])
```

Comment:
Is the copy needed? Should be safe to pass ref[32:] to decryptChunkData as it does not mutate the slice.